Add GD support

mlarash opened the pull request.

Commit:
Add compatibility to copy binary from storage location to boot location

lippit commented:
Please clean up the commit history!
You could probably squash all commits together into one, or only keep multiple commits if each of them works and makes sense by itself.

Also make sure to write a descriptive commit message, look into earlier messages as examples.

Is the MR Gigadevice specific, or does it still run STM? Please also update the readme about the new features and what their purpose is.

lippit commented on src/Bootloader_stm.cpp:
```diff
@@ -24,16 +24,21 @@
 {
+    /* enable watchdog if necessary */
+    if (enableWatchdog) {
+        system.enableWatchdog();
```
Linking fails for me when executing `ninja test` with this error:
/home/georg/okra/okra-bootloader/build/../src/Bootloader_stm.cpp:31: undefined reference to `System::enableWatchdog()'

mlarash replied:
hmm does it still fail for you?

lippit replied:
Nope, compiles now!

lippit commented on meson_options.txt:
```diff
@@ -0,0 +1 @@
+option('GIGADEVICE', type : 'feature', description : 'Enables gigadevice build', yield : true)
```
Is this really Gigadevice-specific and does not run on STM? If also works on STM, I'd prefer keeping it generic and call it "COPYBINARY" or something.

mlarash replied:
Yeah, it will run fine on STM as long as there is space for the new app B address

lippit commented on src/Bootloader_gd.cpp:
```diff
@@ -0,0 +1,116 @@
+/*
```
This file is very similar to Bootloader_stm.cpp, only very few lines are different. I'd prefer having only one file for both architectures.

lippit commented on src/FlashIAP.cpp:
```diff
@@ -0,0 +1,78 @@
+    return true;
+}
+
+bool FlashIAP::isAligned(uint32_t number, uint32_t alignment)
```
This function is not used anywhere, can be removed.

lippit commented on src/FlashIAP.cpp:
```diff
@@ -0,0 +1,78 @@
+    return (number % alignment) == 0;
+}
+
+bool FlashIAP::isAlignedToPage(uint32_t address, uint32_t size)
```
Also unused.

lippit commented on src/FlashIAP.h:
```diff
@@ -0,0 +1,30 @@
+
+    void copyFlashBlock(uint32_t sourceAddress, uint32_t destinationAddress, int32_t size);
+
+    // TODO: confirm thread safety
```
The bootloader does not use threads at all.

lippit commented on src/Config.h:
```diff
@@ -1,11 +1,11 @@
 
 /* Length of the bootloader name string (including \0 termination) */
-const uint8_t BOOTLADER_NAME_LENGTH = 16;
+const uint8_t BOOTLOADER_NAME_LENGTH = 16;
```
Good catch ;-)

lippit commented on src/FlashIAP.cpp:
```diff
@@ -0,0 +1,78 @@
+#include "FlashIAP.h"
```
Now we perform flash operations in both System_stm32.cpp and FlashIAP.cpp, which is a bit inconsistent. I'd prefer having all system dependent flash operations in one place. 

Option 1: Do everything in System_stm32.cpp and scratch FlashIAP.cpp, 
Option 2: Rename your FlashIAP.cpp to FlashIAP_stm32.cpp and move all flash operations to it. Make sure to also add a system-independent FlashIAP_native.cpp to make sure the unit tests can compile, so that the bootlogic test can compile natively.

mlarash replied:
cool, I'll move everything into system then

lippit commented on test/bootlogictest_gd.cpp:
```diff
@@ -0,0 +1,227 @@
+#include "CppUTest/TestHarness.h"
```
Again, I don't see the point of having two test files, because the logic under test is exactly the same. In contrast, a test of the actual copy function is missing. Another useful test would be to check if you didn't accidentally overwrite something outside the target app's range.

mlarash replied:
Okay that makes sense, I've condensed them into a single file. How would you test the copy function without an actual large block of flash to copy? Would you just check that the last programmed address does not extend beyond the expected range?

lippit replied:
What I usually do in such a case:
1. Clear the entire memory (fill it with zeros)
2. Fill your source block with random data (rand() with a known seed or something similar)
3. Run the copy function
4. Check that source and destination blocks are identical now
5. Check that all other memory locations are still zero

lippit replied:
So after reviewing again, I noticed you implemented the actual copy function in a platform dependent way (which makes perfect sense), so there is probably no point in writing native unit tests. It's a rather simple function, I'm fine with not having a unit test for it. Up to you if you want to write one or just leave it as it is.

lippit reviewed (changes requested):
Summary of my review:
There is a lot of code duplication now, which should be avoided under all circumstances. Please keep a single Bootloader.cpp file instead of the two almost identical Bootloader_gd.cpp and Bootloader_stm.cpp

mlarash commented:
> Please clean up the commit history!
> You could probably squash all commits together into one, or only keep multiple commits if each of them works and makes sense by itself.
> 
> Also make sure to write a descriptive commit message, look into earlier messages as examples.
> 
> Is the MR Gigadevice specific, or does it still run STM? Please also update the readme about the new features and what their purpose is.

@lippit I've removed the duplicate code and updated the readme with the new build option. Could you take another look when you have a chance? Thanks for the thorough review.

Regarding the commit message, I'll condense all the commits and write a descriptive message once it's good to go

lippit commented on src/System_stm32f1.cpp:
```diff
@@ -112,6 +115,83 @@ void System::executeFromAddress(uint32_t bootAddress)
+void System::readFlash(uint32_t address, uint8_t* data, int32_t size)
+{
+    uint8_t* src = (uint8_t*)address;
+    for (int i = 0; i < size; i++) {
```
@mlarash Pushed a change to use a simple loop instead of memcpy(). This is probably slower, but reduces binary size by ~200 bytes. Can you check if it still works for you?

mlarash replied:
yep still works, thanks!

mlarash commented:
@lippit I've cleaned up the commit history and tested it, seems like it's good to go with your change.
